CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h

```diff
@@ -0,0 +1,75 @@
+#ifndef CondFormats_SiPixelObjects_SiPixelGainForHLTonGPU_h
+#define CondFormats_SiPixelObjects_SiPixelGainForHLTonGPU_h
+
+#include <cstdint>
+#include <cstdio>
+#include <tuple>
+
+// including <cuda_runtime.h> would pull in the dependency on all of CUDA
+// instead, just define away the CUDA specific attributes to keep GCC happy
+#ifndef __host__
+#define __host__
+#endif
```

Comment:
Does this mean that all client code has to `#include <cuda_runtime.h>` before including this header? Could the `cuda_runtime.h` just be included here? If not e.g. because of `CondFormats`, should we look into moving this file elsewhere?

Reply:
> Does this mean that all client code has to `#include <cuda_runtime.h>` before including this header?

I think when compiling with `nvcc` that is done automatically by the compiler.

> Could the `cuda_runtime.h` just be included here?

That would indeed be cleaner - but it requires adding the dependency on CUDA to the BuildFile.

If you are worried that CUDA code ends up including this without the proper definition of `__host__`, `__device__`, *etc.*, we could protect it inside
```c++
#ifndef __CUDACC__
...
#endif  // __CUDACC__
```

Reply:
I agree this could indeed work in practice (when compiled with `gcc` they indeed should be "empty", and with `nvcc` something should fail, at runtime at the latest, if `nvcc` would not provide them). Nevertheless I have an uneasy feeling about these definitions. I suppose a typical technique would be to define a macro along `HOST_DEVICE` instead, on the other hand, if we foresee a migration to a portability technology within a year or so, I'm not sure if it is worth of the effort.

Reply:
As a quick test, I checked what `__host__` and `__device__` are defined to, when compiling with GCC and including `cuda_runtime.h`, with
```bash
g++ -I /usr/local/cuda/include -include cuda_runtime.h -dM -E -x c++ -
```

What I got is
```c++
#define __host__ __location__(host)
#define __device__ __location__(device)
#define __location__(a) __annotate__(a)
#define __annotate__(a) 
```
*i.e.* they and up being defined to nothing.
So defining them to be empty when compiling with plain GCC should be consistent.

While doing that I realised that `cuda_runtime.h` does not define `__restrict__` - in fact, GCC is supposed to understand it: https://gcc.gnu.org/onlinedocs/gcc/Restricted-Pointers.html .
Do you think it's worth un-defining it and passing it as-is to GCC ?

Reply:
> Do you think it's worth un-defining it and passing it as-is to GCC ?

I think it would be better to pass `__restrict__` to gcc. I see it being used in some CPU-only code already, so it should not cause a problem.

Reply:
By the way, I've added back the `#ifndef __host__` *etc.* check, to avoid a re-definition of the macro in the cases where we are building with GCC *and* we have already included `cuda_runtime.h` .